In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Importar las librerías con el código desarrollado
import sys
sys.path.insert(0,'/content/drive/MyDrive/Colab Notebooks/MLOPS')

In [3]:
# Librerías a usar en el proyecto
from download_data import get_data
import pandas as pd
import plotly.express as px
from statistics import mean

Mounted at /content/drive
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


In [4]:
# Obtención de los archivos
news_true, news_false = get_data()

# **1. Análisis del archivo noticias reales**
De acuerdo con el análisis exploratorio de los datos se encuentra:
1. Total de noticias: 21417
2. Noticias en el año 2016: 4716
3. Noticias en el año 2017: 16701
4. Las noticias se encuentran asociadas a dos temas: 'politicsNews' y 'worldnews'. 
5. Noticias que tratan de politicsNews : 11272
6. Noticias que tratan de worldnews: 10145

In [5]:
# Conteo del número de registros
news_true.count()

title      21417
text       21417
subject    21417
date       21417
dtype: int64

In [6]:
# Información del dataframe
news_true.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB


In [7]:
# Convertir el formato de la columna date a datetime
news_true['date'] = pd.to_datetime(news_true['date'], infer_datetime_format=True)

In [8]:
# Extraer el año de la fecha
news_true['anno'] = news_true['date'].dt.year

In [9]:
# Validación de la conversión del formato y creación de los nuevos elementos 
news_true.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   title    21417 non-null  object        
 1   text     21417 non-null  object        
 2   subject  21417 non-null  object        
 3   date     21417 non-null  datetime64[ns]
 4   anno     21417 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 836.7+ KB


In [10]:
# Número de noticias por año
dfg=news_true.groupby('anno').count().reset_index()
fig = px.bar(dfg, x='anno', y = 'title', title = 'Número de noticias reales por año', color = 'title', barmode = 'stack', text_auto='.2s',
              labels={'anno':'Año de la noticia','title':'Número de noticias'})
fig.show()

In [11]:
# Análisis del subject de las noticias
print(news_true['subject'].unique())

['politicsNews' 'worldnews']


In [12]:
# Número de noticias por tópico
dfg=news_true.groupby('subject').count().reset_index()
fig = px.bar(dfg, x='subject', y = 'title', title = 'Número de noticias reales por tópico', color = 'title', barmode = 'stack',
              labels={'subject':'Tópico de la noticia','title':'Número de noticias'})
fig.show()

In [ ]:
# Análisis del contenido de las noticias tanto en el título como en el texto
news_true.iloc[15001, 1]

'PANAMA CITY (Reuters) - Panama s President Juan Carlos Varela will travel to China to meet with his counterpart Xi Jinping on Friday, the first official visit by a Panamanian leader to the Asian country  coming five months after the nations established diplomatic relations. The bilateral meeting will serve to  establish a new economic, trade, tourist and diplomatic outlook for the country, leading to more than a dozen agreements that will be signed  between nations, Panama s Foreign Ministry said in a statement. In June, Panama upgraded its commercial ties with China and established full diplomatic links with the second most important customer of its key shipping canal, giving Beijing a major victory as it broke formal relations with Taiwan. The Foreign Ministry said the agreements will provide the basis for attracting investments and Chinese innovation to Panama, and help boost Panamanian exports, tourism and the use of the Panama canal. Varela will inaugurate Panama s first embassy 

In [19]:
# Cálculo de la extensión del título (title) y contenido (text) de las noticias reales
news_true['len_title'] = (news_true['title'].str.strip()).str.len()
news_true['len_text'] = (news_true['text'].str.strip()).str.len()

In [39]:
# Análisis de la extensión del título (title) y contenido (text) de las noticias reales
min_title = min(news_true['len_title'])
max_title = max(news_true['len_title'])
ave_title = mean(news_true['len_title'])
min_text = min(news_true['len_text'])
max_text = max(news_true['len_text'])
ave_text = mean(news_true['len_text'])

In [40]:
print(f"Tamaño Máximo del título: {max_title} caracteres. Tamaño Mínimo del título: {min_title} caracteres, Tamaño promedio del título: { round(ave_title)} caracteres")
print(f"Tamaño Máximo del contenido: {max_text} caracteres. Tamaño Mínimo del contenido: {min_text} caracteres, Tamaño promedio del contenido: { round(ave_text)} caracteres")

Tamaño Máximo del título: 133 caracteres. Tamaño Mínimo del título: 26 caracteres, Tamaño promedio del título: 65 caracteres
Tamaño Máximo del contenido: 29780 caracteres. Tamaño Mínimo del contenido: 0 caracteres, Tamaño promedio del contenido: 2382 caracteres


In [41]:
# Se muestran los registros con longitud 1 en el campo text
news_true_short = news_true[news_true['len_text'] == 0]

In [42]:
# solo un registro con longutid 0
news_true_short 

title text       subject       date  anno  \
8970  Graphic: Supreme Court roundup       politicsNews 2016-06-16  2016   

      len_title  len_text  
8970         30         0

In [43]:
# Se crea un nuevo dataframe eliminando el registro sin información en el campo text
news_true_mod = news_true[news_true['len_text'] > 0]

In [44]:
# Se calcula de nuevo los valores min, max y promedio en longitud 
min_text = min(news_true_mod['len_text'])
max_text = max(news_true_mod['len_text'])
ave_text = mean(news_true_mod['len_text'])

In [45]:
print(f"Tamaño Máximo del contenido: {max_text} caracteres. Tamaño mínimo del contenido: {min_text} caracteres, Tamaño promedio del contenido: { round(ave_text)} caracteres")

Tamaño Máximo del contenido: 29780 caracteres. Tamaño mínimo del contenido: 151 caracteres, Tamaño promedio del contenido: 2382 caracteres


# **2. Análisis del archivo noticias falsas**
De acuerdo con el análisis exploratorio de los datos se encuentra:
1. Total de noticias: 23481.
2. El campo 'date' presenta diferentes formatos e incluso datos que no corresponden a la fecha, sin embargo dado el problema que se plantea y como resultado de análisis exploratorio no se considera información relevante para el desarrollo del objetivo del proyecto.
4. Las noticias se encuentran asociadas a dos temas: 'News', 'politics', 'Government News', 'left-news', 'US_News' y 'Middle-east.
5. Noticias que tratan de politics : 6841.
6. Noticias que tratan de News: 9050.
7. Noticias que tratan de Government News: 1570.
8. Noticias que tratan de left-News: 4459.
9. Noticias que tratan de US_News: 783.
10. Noticias que tratan de Middle-east: 778.

In [29]:
# Conteo del número de registros
news_false.count()

title      23481
text       23481
subject    23481
date       23481
dtype: int64

In [ ]:
# Información del dataframe
news_false.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


In [30]:
# Análisis del subject de las noticias
print(news_false['subject'].unique())

['News' 'politics' 'Government News' 'left-news' 'US_News' 'Middle-east']


In [61]:
# Número de noticias por tópico
dfg=news_false.groupby('subject').count().reset_index()
fig = px.bar(dfg, x='subject', y = 'title', title = 'Número de noticias falsas por tópico', color = 'title', barmode = 'stack',text_auto='.2s',
              labels={'subject':'Tópico de la noticia','title':'Número de noticias'})
fig.show()

In [32]:
# Análisis exploratorio del campo date, al encontrarse que no todas las filas tienen una fecha en esa columna
serie = news_false.iloc[9350:9400,3]
print(serie)

9350                                         Nov 26, 2017
9351                                         Nov 26, 2017
9352                                         Nov 26, 2017
9353                                         Nov 25, 2017
9354                                         Nov 25, 2017
9355                                         Nov 25, 2017
9356                                         Nov 25, 2017
9357                                         Nov 25, 2017
9358    https://100percentfedup.com/served-roy-moore-v...
9359                                         Nov 25, 2017
9360                                         Nov 24, 2017
9361                                         Nov 24, 2017
9362                                         Nov 24, 2017
9363                                         Nov 23, 2017
9364                                         Nov 23, 2017
9365                                         Nov 22, 2017
9366                                         Nov 22, 2017
9367          

In [33]:
# Ejemplo de uno de los registros y lo que aparece registrado en ese campo
serie9358 = news_false.iloc[9358,:]
print(serie9358[3])

https://100percentfedup.com/served-roy-moore-vietnamletter-veteran-sets-record-straight-honorable-decent-respectable-patriotic-commander-soldier/


In [34]:
# Convertir el formato de la columna date a datetime
news_false['date_new'] = pd.to_datetime(news_false['date'], infer_datetime_format=True, errors='coerce')

In [35]:
# Exploración del campo creado para la fecha
news_false['date_new'].unique()

array(['2017-12-31T00:00:00.000000000', '2017-12-30T00:00:00.000000000',
       '2017-12-29T00:00:00.000000000', '2017-12-25T00:00:00.000000000',
       '2017-12-23T00:00:00.000000000', '2017-12-22T00:00:00.000000000',
       '2017-12-21T00:00:00.000000000', '2017-12-20T00:00:00.000000000',
       '2017-12-19T00:00:00.000000000', '2017-12-17T00:00:00.000000000',
       '2017-12-16T00:00:00.000000000', '2017-12-13T00:00:00.000000000',
       '2017-12-12T00:00:00.000000000', '2017-12-11T00:00:00.000000000',
       '2017-12-10T00:00:00.000000000', '2017-12-09T00:00:00.000000000',
       '2017-12-07T00:00:00.000000000', '2017-12-06T00:00:00.000000000',
       '2017-12-05T00:00:00.000000000', '2017-12-04T00:00:00.000000000',
       '2017-12-03T00:00:00.000000000', '2017-12-02T00:00:00.000000000',
       '2017-12-01T00:00:00.000000000', '2017-11-30T00:00:00.000000000',
       '2017-11-29T00:00:00.000000000', '2017-11-28T00:00:00.000000000',
       '2017-11-27T00:00:00.000000000', '2017-11-26

In [36]:
# Cálculo de la extensión del título (title) y contenido (text) de las noticias reales
news_false['len_title'] = (news_false['title'].str.strip()).str.len()
news_false['len_text'] = (news_false['text'].str.strip()).str.len()

In [46]:
# Análisis de la extensión del título (title) y contenido (text) de las noticias reales
min_title_false = min(news_false['len_title'])
max_title_false = max(news_false['len_title'])
ave_title_false = mean(news_false['len_title'])
min_text_false = min(news_false['len_text'])
max_text_false = max(news_false['len_text'])
ave_text_false = mean(news_false['len_text'])

In [47]:
print(f"Tamaño Máximo del título: {max_title_false} caracteres. Tamaño Mínimo del título: {min_title_false} caracteres, Tamaño promedio del título: { round(ave_title_false)} caracteres")
print(f"Tamaño Máximo del contenido: {max_text_false} caracteres. Tamaño Mínimo del contenido: {min_text_false} caracteres, Tamaño promedio del contenido: { round(ave_text_false)} caracteres")

Tamaño Máximo del título: 279 caracteres. Tamaño Mínimo del título: 8 caracteres, Tamaño promedio del título: 94 caracteres
Tamaño Máximo del contenido: 51793 caracteres. Tamaño Mínimo del contenido: 0 caracteres, Tamaño promedio del contenido: 2547 caracteres


In [59]:
# Registros sin información en el campo texto
reg_sininfo = news_false[news_false['len_text'] == 0 ]
reg_sininfo.count()

title        630
text         630
subject      630
date         630
date_new      39
len_title    630
len_text     630
dtype: int64

In [52]:
# Se crea un nuevo dataframe eliminando el registro sin información en el campo text
news_false_mod = news_false[news_false['len_text'] > 0]

In [60]:
# Número de reistro en el dataframa de noticias falsa modificados
news_false_mod.count()

title        22851
text         22851
subject      22851
date         22851
date_new     11829
len_title    22851
len_text     22851
dtype: int64

In [53]:
# Se calcula de nuevo los valores min, max y promedio en longitud 
min_text_false = min(news_false_mod['len_text'])
max_text_false = max(news_false_mod['len_text'])
ave_text_false = mean(news_false_mod['len_text'])

In [54]:
print(f"Tamaño Máximo del contenido: {max_text_false} caracteres. Tamaño mínimo del contenido: {min_text_false} caracteres, Tamaño promedio del contenido: { round(ave_text_false)} caracteres")

Tamaño Máximo del contenido: 51793 caracteres. Tamaño mínimo del contenido: 5 caracteres, Tamaño promedio del contenido: 2617 caracteres
